### ISSUE TITLE
Question about ifc unit.

### ISSUE THREAD

**Issue by YauronBis** [xBimTeam/XbimEssentials#189]
Hello.
Today I am in a process of understanding quite basic. I am trying to match the unit of measure declaration lines (IFCSIUNIT) that precede the project declaration (IfcProjet) and the units of measure used according to the IFC object. Indeed, I do not find in the EXPRESS schemas an indication that this unit of measure is used by such an IFC object, so the values ​​I get are without unity. I would like to understand this link to reproduce the reasoning in my program to automatically accompany the numerical value of the unit that corresponds to it.

**Comment by YauronBis** [comment]
or in other words, how does one determine the unit of measure of an IfcMeasureValue object according to IfcSiUnit declarations?

**Comment by martin1cerny** [comment]
You have to create a map between `IfcMeasureValue` types and applicable unit definitions in IFC.

**Comment by YauronBis** [comment]
Ok, so something like KeyValuePair(of IfcValue_ObjectName as string, UnitValue AS IfcUnitEnum) to link on IfcSiUnit on IfcUnitEnum value.

**Comment by martin1cerny** [comment]
Yes, exactly.

**Comment by YauronBis** [comment]
i'm on my way, but, does this means xBim do not have a similar [map ](http://dcom.arch.gatech.edu/aci/cic/schema/ifcmeasureresource/content.htm) within ?

**Comment by martin1cerny** [comment]
No, it doesn't. We have mapping for length units and their conversions so we can handle geometry correctly and there is some code to present basic quantities (length, area, volume,...) in the UI but it is certainly not complete.

**Comment by YauronBis** [comment]
I did it last week.
In my own NameSpace, called IFCX :
- I add "Unity" class that resume IFC complex class organisation in one (Monetary, Named, Derived).
- I add a sub class in Meta Dictionary as MeasureObject Type + Meta relation between MeasureObject type and EnumValue Type, to be able to associate an IFC enumValue from aggregate IFCUnitEnum + IfcDerivedUnitEnum, to one MeasureObject definition.
- I Add some Meta description in my Meta Dictionary for a new IFC attribute definition AS "Symbol". Then Unity set is own symbol as much he can, and when it can't it allow user to define it and store user Symbol.
- Some code find the right Symbol attribute from IFC original Object type, making the Property Abstract in Unity class.
So each MeasureObject know is unity enum unity reference, and ask from IFCX imported whole objects, The unity concern by the same unity  enum value, and deduce symbol.
![unite de mesure](https://user-images.githubusercontent.com/39408265/43144334-28838bc4-8f5d-11e8-9d94-d8d99d5cbeb8.PNG)

**Comment by martin1cerny** [comment]
Good work! Will you be willing to contribute to xBIM project with your code (or the basic map generation between `IfcValue` type and `IfcUnit`)? I'm sure other members of the community might appreciate this as well.

**Comment by YauronBis** [comment]
I really wish i could, but there is two problems.

First, in France, my boss is the owner of all my work, so it would be a professional error to give it without his approval. And be sure, he will never accept it. All i can say publicly about it, is about general concepts.

Secondly, as weird as it may seem, for the time being, I do not program directly with xbim. Let me explain. From the point of view of xBim, objects and properties are those that are described by BuildingSmart. A change on their part, must generate a change in the code of xBIM. It is a chain of incidence that can be related to a constraint. A constraint that I do not assume.

From my point of view, things must necessarily be simpler, (hence my IFCX and DICX format) and less constraining: there are objects, polymorphism, enumerators, enumerated values, and so thus types, objects attributes, block of properties (PSET), properties, and therefore associations between these different notions. xBim is great content otherwise, clearly, I will not be able to do anything! 

xBim role in my project is to be my translator ! "Here is an IFC file, Here is the restricted list (from the Meta Model database) of objects / properties / ... that I want to recover" by means of a query and a simple algorithm, xBIM query its memory (LINQ ) and extract this information (CallByName), to sum it all up in 2 distinct tables in my IFCX database: a list of IFC objects reference, a list of couple {IFC attributes +IFC values}. The job of my IFCX library (.dll) is to be able to reconstruct this extracted information, in an equally simple and reductive memory format : List(of IFCX_Object), List(of IFCX_Values), both bound to List(of DICX_ElementDefinition), letting IFCX_Object to be able to get back Relative IFCX_Values and Relative DICX information. 

We can say that this approach has the merit of being dynamic, even if, the programming is not really.
We can say, that, IFCX is an emulation of IFC information.
For example, IFC do not define Symbol Attribute for unity definition. I just have to add it in our dictionary, stating that this definition comes from us and not from a specific IFC version, so i can store Symbol for IFCX_unity (inherit IFCX_Object) like for Degree that do not exist anywhere in IFC.

**Comment by martin1cerny** [comment]
That's fine :-) It is just nice when people using xBIM can contribute to community. But is is obviously just an option and your reasons are clear enough. You might be surprised but there are people who work under same conditions as you who contribute their work (with a consent of their employers I assume).

**Comment by YauronBis** [comment]
Yes but I want to send back help as much as possible.. For my solution around IFC units to make sense, it would imply the global revelation of my framework (so far 2 years of work without return on investment), For fear of not being clear enough in my previous explanation: it consists in summarize [the complexity of the IFC](http://www.buildingsmart-tech.org/ifc/IFC2x3/TC1/html/ifcmeasureresource/diagrams/diagram_0001.htm) in a single IFCX_Unity class ![ifcx_unity](https://user-images.githubusercontent.com/39408265/43255376-f28c3f52-90c9-11e8-916e-7836acc05094.PNG), to articulate it with those of the IFCX_MeasureObject measurement objects. Articulation is only possible if a Measurement Object shares a common sense with the units: this information, as we stated in our exchange, is absent from the IFC and xBIM. to be consistent with my programming, I introduced this in the form of an injection of descriptive data rather than code.

I also added that my framework is the only technical solution that seemed to lead to an automatic multilingual translation solution, which for French people is an important thing.
